File: 09_DataStructure_Algorithm/0518_day51/소수 찾기/소수 찾기.py
from itertools import permutations

def solution(numbers):
    answer = set()
    
    def check_prime(n):
        if n < 2:
            return False
        
        for i in range(2, int(n ** (1/2)) + 1):
            if n % i == 0:
                return False
        return True
    
    for i in range(len(numbers)):
        for permu in permutations(numbers, i + 1):
            num = int(''.join(permu))
            if check_prime(num):
                answer.add(num)
    
    return len(answer)

File: 09_DataStructure_Algorithm/0518_day51/소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    public int solution(String numbers) {
        int answer = 0;
        
        // nP1, nP2, ... nPn : 모든 경우의 수 나열
        Set<Integer> candidates = new HashSet<>();
        permute(numbers, new boolean[numbers.length()], new StringBuilder(), candidates);
        
        for (int num : candidates) {
            if (isPrime(num)) answer++;
        }
        
        return answer;
    }
    
    void permute(String numbers, boolean[] visited, StringBuilder cur, Set<Integer> candidates) {
        if (cur.length() > 0) {
            candidates.add(Integer.parseInt(cur.toString()));
        }
        
        for (int i = 0; i < numbers.length(); i++) {
            if (!visited[i]) {
                cur.append(numbers.charAt(i));
                visited[i] = true;
                permute(numbers, visited, cur, candidates);
                visited[i] = false;
                cur.deleteCharAt(cur.length() - 1);
            }
        }
    }
    
    boolean isPrime(int num) {
        if (num < 2) return false;
        
        for (int i = 2; i * i <= num; i++) {
            if (num % i == 0) return false;
        }
        return true;
    }
}